### PR TITLE
refactor!: Streamline and rename algebra types

### PR DESCRIPTION
**Pull request by msmk0**
BREAKING CHANGE: The number of generic algebra types with configurable size is reduced, existing types are renamed or have changed template arguments. All algebra types with variable scalar and additional unused types have been removed. Only a minimal set of algebra types with a fixed underlying scalar type `ActsScalar` and configurable size is provided.

Types prefixed with `Acts...` always use `ActsScalar` as the underlying scalar type. Different types exists to allow variable size at compile- or runtime:

```cpp
using ActsVector = Eigen::Matrix<ActsScalar, kSize, 1>;
using ActsMatrix = Eigen::Matrix<ActsScalar, kRows, kCols>;
using ActsSymMatrix = Eigen::Matrix<ActsScalar, kSize, kSize>;
using ActsDynamicVector = Eigen::Matrix<ActsScalar, Eigen::Dynamic, 1>;
using ActsDynamicMatrix = Eigen::Matrix<ActsScalar, Eigen::Dynamic, Eigen::Dynamic>;
```

The fixed-sized types `Vector{2,3,4}D`, `SymMatrix{2,3,4}D`, ... now also use the `ActsScalar` scalar type. Their names are left as-is. Renaming them to `Vector{2,3,4}`, ... without the `D` suffix is a simple find/replace and will be handled in a separate PR.

All row-vector types have been removed. Having both (column-)vector and row-vectors is confusing because they behave very differently, e.g. for matrix-vector multiplications. For the only existing use case, the free parameter path derivatives, a distinct `FreeParametersToMatrix` type is introduced (named similar to `FreeToBoundMatrix` and `BoundToFreeMatrix`).

For the cases where a scalar type different from `ActsScalar` is needed, `Eigen::Matrix` is used directly; this is the case e.g. in various utility tools that are not part of the public interface. To avoid having `Eigen::Matrix` in public interfaces, appropriate typedefs are introduced locally and used consistently:

- `Material::ParametersVector = Eigen::Matrix<float, 5, 1>` in the material module.
- `MainGridVector = Eigen::Matrix<float, ..., 1>` and `TrackGridVector = Eigen::Matrix<float, ..., 1>` in various algorithms in the vertexing module.

The `{Bound,Free,Alignment}Scalar` types are removed and replaced with `ActsScalar`.

The `LocalCartesianToBoundLocalMatrix` is removed and replaced with `ActsMatrix<2,3>`.

Replaces and closes #595.

### CODE REVIEW

**Comment by msmk0** [comment]
There are a two further changes that I think we should consider:

- To remove the separate `BoundScalar`, `FreeScalar`, and `AlignmentScalar` types and use `ActsScalar` everywhere instead. The default setup is already such that they all map to `ActsScalar`. Removing the separate types will also allow us to remove most remaining explicit `Eigen::Matrix<...>` uses as they are currently required to define types based on `{Bound,Free,Alignment}Scalar`.
- Replace `PositionToBoundLocalMatrix` directly with `ActsMatrix<3,2>`. This projection from 3d into 2d local space is not specific to bound parameters and does not warant a separate type in my opinion. With the streamlined algebra types

**Comment by msmk0** [comment]
As discussed this morning, I will go ahead with the additional proposed changes.

**Comment by msmk0** [comment]
@asalzburger @paulgessinger This is ready to be reviewed. I also put up #618 that removes the `D` suffix from the fixed-size coordinates types.

File: Core/include/Acts/Definitions/Alignment.hpp [hunk 1]
```diff
@@ -32,17 +32,11 @@ enum AlignmentIndices : unsigned int {
   eAlignmentSize,
 };
 
-/// Underlying fundamental Scalar type for alignment parameters.
-using AlignmentScalar = ActsScalar;
```
Comment (asalzburger) [inline Core/include/Acts/Definitions/Alignment.hpp:36]: We might have to check if we need a dedicated scalar setting if we want to use auto-diff for the alignment?

File: Core/include/Acts/EventData/MultiTrajectory.hpp [hunk 2]
```diff
@@ -95,7 +95,7 @@ struct Types {
     Flags = Eigen::ColMajor | Eigen::AutoAlign,
     SizeIncrement = 8,
   };
-  using Scalar = double;
+  using Scalar = ActsScalar;
```
Comment (asalzburger) [inline Core/include/Acts/EventData/MultiTrajectory.hpp:98]: Oh sorry, I had missed that.

File: Core/include/Acts/EventData/detail/PrintParameters.hpp [hunk 3]
```diff
@@ -42,24 +42,22 @@ void printFreeParameters(std::ostream& os, const FreeVector& params,
 /// @param os The output stream
 /// @param size Size of the measurement space
 /// @param indices Which parameters are measured, must contain size elements
-/// @param params Parameters vector
-/// @param cov Optional Covariance matrix
+/// @param params Parameters vector data, must contain size elements
+/// @param cov Optional Covariance matrix data, must contain sizexsize elements
```
Comment (asalzburger) [inline Core/include/Acts/EventData/detail/PrintParameters.hpp:46]: sizexsize?

File: Core/include/Acts/Surfaces/ConeSurface.hpp [hunk 4]
```diff
@@ -219,7 +219,7 @@ class ConeSurface : public Surface {
   ///
   /// @return Derivative of bound local position w.r.t. position in local 3D
   /// cartesian coordinates
-  LocalCartesianToBoundLocalMatrix localCartesianToBoundLocalDerivative(
+  ActsMatrix<2, 3> localCartesianToBoundLocalDerivative(
```
Comment (asalzburger) [inline Core/include/Acts/Surfaces/ConeSurface.hpp:222]: Why is this name removed?

File: Core/include/Acts/Utilities/Ray.hpp [hunk 5]
```diff
@@ -23,7 +23,7 @@ class Ray {
   /// Re expose the value type
   using value_type = value_t;
   /// Vertex type based on the value type and dimension
-  using VertexType = ActsVector<value_t, DIM>;
+  using VertexType = Eigen::Matrix<value_t, DIM, 1>;
   /// Vertex array type corresponding to the vertex type
   using vertex_array_type = Eigen::Array<value_t, DIM, 1>;
   /// Associated transform type
```
Comment (asalzburger) [inline Core/include/Acts/Utilities/Ray.hpp:29]: We have ray and frustum only in the unit tests as float types btw.

File: Core/include/Acts/Visualization/ObjVisualization3D.hpp [hunk 6]
```diff
@@ -35,7 +35,7 @@ class ObjVisualization3D : public IVisualization3D {
   using ValueType = T;
 
   /// Type of a vertex based on the value type
-  using VertexType = ActsVector<ValueType, 3>;
+  using VertexType = Eigen::Matrix<ValueType, 3, 1>;
 
   /// Type of a line
   using LineType = std::pair<size_t, size_t>;
```
Comment (asalzburger) [inline Core/include/Acts/Visualization/ObjVisualization3D.hpp:41]: Same as for frustum/ray - exists only in unit tests, so we could probably remove this customisation.

File: Tests/UnitTests/Core/Utilities/RayTest.cpp [hunk 7]
```diff
@@ -22,7 +22,7 @@ BOOST_AUTO_TEST_SUITE(Utilities)
 BOOST_AUTO_TEST_CASE(ray_construction) {
   // 2D
```
Comment (asalzburger) [inline Tests/UnitTests/Core/Utilities/RayTest.cpp:23]: As written above ...

**Review by asalzburger — APPROVED** [review]
I really like this PR - I have a few comments worth considering, but in general I approve.

File: Core/include/Acts/Definitions/Algebra.hpp [hunk 8]
```diff
@@ -22,71 +22,75 @@
...
 ///
 /// @{
 
 // coordinate vectors
-using Vector2D = ActsVector<ActsScalar, 2>;
-using Vector3D = ActsVector<ActsScalar, 3>;
-using Vector4D = ActsVector<ActsScalar, 4>;
+using Vector2D = ActsVector<2>;
+using Vector3D = ActsVector<3>;
+using Vector4D = ActsVector<4>;
```
Comment (paulgessinger) [inline Core/include/Acts/Definitions/Algebra.hpp:80]: So these are just aliases?

Comment (msmk0) [reply Core/include/Acts/Definitions/Algebra.hpp:80]: Yes.

File: Core/include/Acts/Definitions/Algebra.hpp [hunk 9]
```diff
@@ -22,71 +22,75 @@
...
+/// Defaults to `double` but can be customized by the user.
 #ifdef ACTS_CUSTOM_SCALARTYPE
 using ActsScalar = ACTS_CUSTOM_SCALARTYPE;
 #else
 using ActsScalar = double;
 #endif
 
-template <typename T, unsigned int rows, unsigned int cols>
-using ActsMatrix = Eigen::Matrix<T, rows, cols>;
+template <unsigned int kSize>
```
Comment (paulgessinger) [inline Core/include/Acts/Definitions/Algebra.hpp:53]: As per your code guidelines, shouldn't this be something like `vector_size_t`?

File: Core/include/Acts/Definitions/Algebra.hpp (hunk 8 above)
Comment (paulgessinger) [reply Core/include/Acts/Definitions/Algebra.hpp:80]: Then I don't quite understand why we don't just stick to the names we use already.

File: Core/include/Acts/Definitions/Algebra.hpp (hunk 9 above)
Comment (msmk0) [reply Core/include/Acts/Definitions/Algebra.hpp:53]: We have always applied *our* [naming guideline N.5](https://acts.readthedocs.io/en/latest/codeguide.html#n-5-constant-values-use-kcamelcase), i.e. template value parameters are treated as constants. [N.7](https://acts.readthedocs.io/en/latest/codeguide.html#n-7-template-parameter-types-use-snake-case-t) only applies to template type parameters.

I think we previously had a discussion on the same topic and decided to follow the [Google Styleguide (bottom paragraph)](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules) on this. Thus the `k...` naming style.

File: Core/include/Acts/Definitions/Algebra.hpp (hunk 8 above)
Comment (msmk0) [reply Core/include/Acts/Definitions/Algebra.hpp:80]: I do not quite understand. The `Vector{2,3,4}D` et al. types are the ones that we currently use?

File: Core/include/Acts/Definitions/Alignment.hpp (hunk 1 above)
Comment (msmk0) [reply Core/include/Acts/Definitions/Alignment.hpp:36]: I expect that using auto-diff for alignment will be handled similar to its use in the stepper code, i.e. the implementation is templated on the scalar type which defaults to `ActsScalar`. This is orthogonal to having a separate `AlignmentScalar` typedef.

File: Core/include/Acts/EventData/detail/PrintParameters.hpp (hunk 3 above)
Comment (msmk0) [reply Core/include/Acts/EventData/detail/PrintParameters.hpp:46]: Would you prefer size²?

File: Core/include/Acts/Definitions/Algebra.hpp (hunk 9 above)
Comment (paulgessinger) [reply Core/include/Acts/Definitions/Algebra.hpp:53]: You're right. Ok!

File: Core/include/Acts/Definitions/Algebra.hpp (hunk 8 above)
Comment (paulgessinger) [reply Core/include/Acts/Definitions/Algebra.hpp:80]: Hm. I guess this is what you meant renaming `VectorXD` to `ActsVectorX`. Ok nevermind, let's leave this as is.

Comment (msmk0) [reply Core/include/Acts/Definitions/Algebra.hpp:80]: The PR description was not clear enough: this only streamlines the generic vector/matrix types with configurable size, not the fixed-sized types intended for coordinates.

**Pull request merged**